### PR TITLE
Fix PixbufFormat, regen Pixbuf

### PR DESCRIPTION
**Pull request by EPashkin**
Closes https://github.com/gtk-rs/gdk-pixbuf/issues/55
Closes https://github.com/gtk-rs/gdk-pixbuf/issues/56
Closes https://github.com/gtk-rs/gdk-pixbuf/issues/23

Build will fails until https://github.com/gtk-rs/glib/pull/278

cc @GuillaumeGomez, @sdroege

### CODE REVIEW

**Comment by EPashkin** [comment]
Ok, PR is finished,
Not all async functions generated by unknown reason, but we can finish it later.

**Comment by GuillaumeGomez** [comment]
I agree. Once CI is good, it's good for me as well.

**Comment by EPashkin** [comment]
Forgot to check async function, at minimum `new_from_stream_async` need be manual.

**Comment by sdroege** [comment]
Generally looks good, good work :) will check later in detail, not ideal from the phone

**Comment by EPashkin** [comment]
Updated.

**Comment by EPashkin** [comment]
@sdroege Please, look at this if you have time.

**Comment by sdroege** [comment]
I will tomorrow, thanks for the reminder. I missed that something is left to be done here

**Comment by sdroege** [comment]
Also feel free to remind me sooner than ~2 weeks later for such things in the future :)

File: src/auto/pixbuf.rs [hunk 1]
```diff
@@ -75,90 +76,98 @@ impl Pixbuf {
...
+
     fn apply_embedded_orientation(&self) -> Option<Pixbuf>;
 
+    fn composite(&self, dest: &Pixbuf, dest_x: i32, dest_y: i32, dest_width: i32, dest_height: i32, offset_x: f64, offset_y: f64, scale_x: f64, scale_y: f64, interp_type: InterpType, overall_alpha: i32);
+
+    fn composite_color(&self, dest: &Pixbuf, dest_x: i32, dest_y: i32, dest_width: i32, dest_height: i32, offset_x: f64, offset_y: f64, scale_x: f64, scale_y: f64, interp_type: InterpType, overall_alpha: i32, check_x: i32, check_y: i32, check_size: i32, color1: u32, color2: u32);
+
+    fn composite_color_simple(&self, dest_width: i32, dest_height: i32, interp_type: InterpType, overall_alpha: i32, check_size: i32, color1: u32, color2: u32) -> Option<Pixbuf>;
+
+    fn copy(&self) -> Option<Pixbuf>;
```
Comment (sdroege) [inline src/auto/pixbuf.rs]: Can this really fail? And the others returning an Option

File: src/auto/pixbuf.rs [hunk 2]
```diff
@@ -169,38 +178,28 @@ pub trait PixbufExt {
...
-    //fn save_to_streamv<'a, P: IsA</*Ignored*/gio::OutputStream>, Q: Into<Option<&'a gio::Cancellable>>>(&self, stream: &P, type_: &str, option_keys: &[&str], option_values: &[&str], cancellable: Q) -> Result<(), Error>;
+    #[cfg(any(feature = "v2_36", feature = "dox"))]
+    fn save_to_streamv<'a, P: IsA<gio::OutputStream>, Q: Into<Option<&'a gio::Cancellable>>>(&self, stream: &P, type_: &str, option_keys: &[&str], option_values: &[&str], cancellable: Q) -> Result<(), Error>;
 
     //#[cfg(any(feature = "v2_36", feature = "dox"))]
-    //fn save_to_streamv_async<'a, P: IsA</*Ignored*/gio::OutputStream>, Q: Into<Option<&'a gio::Cancellable>>, R: /*Ignored*/gio::AsyncReadyCallback>(&self, stream: &P, type_: &str, option_keys: &[&str], option_values: &[&str], cancellable: Q, callback: R);
+    //fn save_to_streamv_async<'a, P: IsA<gio::OutputStream>, Q: Into<Option<&'a gio::Cancellable>>, R: /*Unimplemented*/gio::AsyncReadyCallback>(&self, stream: &P, type_: &str, option_keys: &[&str], option_values: &[&str], cancellable: Q, callback: R);
 
     fn savev(&self, filename: &str, type_: &str, option_keys: &[&str], option_values: &[&str]) -> Result<(), Error>;
```
Comment (sdroege) [inline src/auto/pixbuf.rs]: This was already weird before, but it seems like this should be taking a `options: &[(&str, &str)]` instead? It's probably unsafe now if the arrays are different length?

File: src/pixbuf.rs [hunk 3]
```diff
@@ -178,10 +163,44 @@ impl Pixbuf {
         }
     }
 
-    pub fn copy(&self) -> Pixbuf {
```
Comment (sdroege) [inline src/pixbuf.rs]: Here it also was not returning an Option.

Maybe take a diff of the API changes here to make sure the API is not regressing without noticing.

File: src/pixbuf.rs [hunk 4]
```diff
@@ -178,10 +163,44 @@ impl Pixbuf {
...
+    pub fn get_file_info(filename: &str) -> Option<(PixbufFormat, i32, i32)> {
```
Comment (sdroege) [inline src/pixbuf.rs]: This should probably be a &Path?

File: src/auto/pixbuf.rs (hunk 1 above)
Comment (EPashkin) [reply src/auto/pixbuf.rs]: Bad, but yes: https://developer.gnome.org/gdk-pixbuf/stable/gdk-pixbuf-Image-Data-in-Memory.html#gdk-pixbuf-copy

Comment (EPashkin) [reply src/auto/pixbuf.rs]: I will check all others later.

Comment (sdroege) [reply src/auto/pixbuf.rs]: Interesting

File: src/auto/pixbuf.rs (hunk 2 above)
Comment (EPashkin) [reply src/auto/pixbuf.rs]: You right. Maybe even `options: &[(AsRef<str>, AsRef<str>)]`

File: src/pixbuf.rs (hunk 4 above)
Comment (EPashkin) [reply src/pixbuf.rs]: Yes, I change it to `AsRef<Path>`

**Comment by EPashkin** [comment]
@sdroege Updated with `AsRef<Path>` for file names and `&[(&str, &str)]` for save options.
Tried `&[(AsRef<str>, AsRef<str>)]`, but it make problem with saving without options.
About return `Option<Pixbuf>`: seems Pixbuf even don't follow glib rules that object constructors can't return null, so all function returning Pixbuf need be optional.
I stuck with https://github.com/rust-lang/rust/issues/47048 and can't build gir, so I can't change this now and propose postpone this change.

**Comment by sdroege** [comment]
:+1:

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**